rust/protocol/src/error.rs

```diff
@@ -11,82 +15,208 @@ use uuid::Uuid;
-    NoSenderKeyState { distribution_id: Uuid },
+    ///
+    /// Raised if a [SenderKeyStore] is unable to locate a [SenderKeyRecord] for a given
+    /// *([ProtocolAddress], [Uuid])* pair in [group_encrypt] or [group_decrypt].
+    NoSenderKeyState {
+        /// Unique identifier for the sender key session.
+        distribution_id: Uuid,
+    },
 
     /// session with {0} not found
+    ///
+    /// Raised if an [IdentityKeyStore] does not contain an entry for a [ProtocolAddress], or
+    /// alternately if a [SessionStore] does not contain a session for a given [ProtocolAddress].
     SessionNotFound(crate::ProtocolAddress),
     /// invalid session: {0}
+    ///
+    /// Raised if a [SessionStore] does not contain a remote identity key to validate.
+    ///
+    /// Similar to [Self::InvalidState].
     InvalidSessionStructure(&'static str),
     /// invalid sender key session with distribution ID {distribution_id}
-    InvalidSenderKeySession { distribution_id: Uuid },
+    ///
+    /// Raised if a [SenderKeyStore] could not load the session for the given ID.
+    ///
+    /// Similar to [Self::NoSenderKeyState].
+    InvalidSenderKeySession {
+        /// Unique identifier for the sender key session.
+        distribution_id: Uuid,
+    },
     /// session for {0} has invalid registration ID {1:X}
+    ///
+    /// Raised if a sealed sender message has a registration id that doesn't map to any
+    /// known session.
     InvalidRegistrationId(crate::ProtocolAddress, u32),
 
     /// message with old counter {0} / {1}
+    ///
+    /// Raised if the same message is decrypted twice so it can be discarded.
     DuplicatedMessage(u32, u32),
     /// invalid {0:?} message: {1}
+    ///
+    /// Raised if a [SignalMessage] could not be decrypted or some field had an unexpected value.
+    ///
+    /// *TODO: what differentiates this from [Self::CiphertextMessageTooShort]?*
     InvalidMessage(crate::CiphertextMessageType, &'static str),
 
     /// error while invoking an ffi callback: {0}
+    ///
+    /// Raised to propagate an error from an FFI callback.
     FfiBindingError(String),
     /// error in method call '{0}': {1}
+    ///
+    /// Raised to propagate an error through to an FFI exception along with a boxed handle.
     ApplicationCallbackError(
         &'static str,
         #[source] Box<dyn std::error::Error + Send + Sync + UnwindSafe + 'static>,
     ),
 
     /// invalid sealed sender message: {0}
+    ///
+    /// Raised if an [crate::sealed_sender::UnidentifiedSenderMessage] could not be
+    /// deserialized successfully.
+    ///
+    /// *TODO: this sounds a lot like [Self::InvalidProtobufEncoding] or [Self::UntrustedIdentity]?*
     InvalidSealedSenderMessage(String),
     /// unknown sealed sender message version {0}
+    ///
+    /// Raised if an version decoded from a [crate::sealed_sender::UnidentifiedSenderMessage]
+    /// was unrecognized.
     UnknownSealedSenderVersion(u8),
```

Comment:
This won't be decoded from UnidentifiedSenderMessage; it's in the outermost envelope.